sdks/java/io/cassandra/src/test/java/org/apache/beam/sdk/io/cassandra/CassandraIOTest.java

```diff
@@ -268,6 +267,36 @@ public void testWrite() {
     }
   }
 
+  @Test
+  public void testWriteWithPasswordDecryption() {
+    ArrayList<Scientist> scientists = buildScientists(NUM_ROWS);
+
+    String sessionWriteUID = "session-write-" + UUID.randomUUID();
+    PasswordDecrypter writePwdDecrypter = Mockito.spy(new TestPasswordDecrypter(sessionWriteUID));
+
+    pipeline
+        .apply(Create.of(scientists))
+        .apply(
+            CassandraIO.<Scientist>write()
+                .withHosts(Arrays.asList(CASSANDRA_HOST))
+                .withPort(CASSANDRA_PORT)
+                .withUsername(CASSANDRA_USERNAME)
+                .withEncryptedPassword(CASSANDRA_ENCRYPTED_PASSWORD)
+                .withPasswordDecrypter(writePwdDecrypter)
+                .withKeyspace(CASSANDRA_KEYSPACE)
+                .withEntity(Scientist.class));
+
+    pipeline.run();
+
+    List<Row> results = getRows();
+    assertEquals(NUM_ROWS, results.size());
+    for (Row row : results) {
+      assertTrue(row.getString("person_name").matches("Name (\\d*)"));
+    }
+
+    assertTrue(1L <= TestPasswordDecrypter.getNbCallsBySession(sessionWriteUID));
```

Comment:
@echauchot All the changes have been made except for the assert method used in tests. I kept the statement "assertTrue" (vs. assertEquals) because the number of calls to the decrypt() method is hazardous, we do not know the exact number of instantiated threads (with local tests I got several results: 11, 3, ...)

Reply:
Sure that depends on the split of the source and direct runner adds some randomization around the number of threads contrary to other runners like spark that have fixed nb of executors based on default parallelism (configured local threads or yarn cluster size ...). So, Fair enough